overte-org/overte
# Improve audio settings

This implements configurable config settings for audio codecs. It's still at an early stage, and only being submitted for feedback and collaboration. Rebased from Vircadia.

TODO list:

* Cleanup code
* API Documentation
* Figure out the final list of audio parameters
* Implement settings in interface
* Fix `Audio.getEncoderFeatures()`


Initial API for audio codecs, client side:

`Audio.getCodec()` -- returns current codec
`Audio.getCodecs()` -- list of known codecs
`Audio.getAllowedCodecs()` -- list of codecs the interface is willing to use. All codecs are accepted if the list is empty. Will check that the names correspond with something in `getCodecs()`, and filter out any unknown entries.
`Audio.setAllowedCodecs()` -- set list of allowed codecs for the interface.
`Audio.getEncoderVBR()` -- is VBR enabled
`Audio.setEncoderVBR(bool)` -- enable/disable Variable Bit Rate
`Audio.getEncoderBitrate()`
`Audio.getEncoderBitrate(int bitrate)` -- Bit rate, in bits/s. 2400 to 512000. Default is 128000 (128kbps)
`Audio.getEncoderComplexity()`
`Audio.getEncoderComplexity(int complexity)` -- How CPU intensive compression is, 0 to 100.
`Audio.getEncoderFEC()` -- is VBR enabled
`Audio.setEncoderFEC(bool)` -- enable/disable Forward Error Correction
`Audio.getEncoderPacketLossPercent()` 
`Audio.setEncoderPacketLossPercent(int percent)` -- how much data loss to allow for

Not working:
`Audio.getEncoderFeatures()`  -- returns a feature list for the current encoder


Testing:

* You can test with yourself by using this console command: `Audio.setServerEcho(true)`. This will bounce audio off the domain at yourself. Microphone has to be on for it to work.

* Use for instance `Audio.setAllowedCodecs(["zlib"])` in the console, and observe the rise in bandwidth usage. 
* Use `Audio.setEncoderBitrate(5000)` to see the effect of 5 kbps audio
* Use `Audio.setEncoderComplexity(1)` to see the effect of less CPU intensive encoding
* Use `Audio.setEncoderFEC(true)` and `Audio.setEncoderPacketLossPercent(10)` to try error correction. Needs a connection with some packet loss.


Feedback would be very welcome -- especially regarding whether the API makes sense, and if there are threading issues anywhere in this.
